From: Gerald Schaefer <geraldsc@de.ibm.com>
Subject: IBM Codestream linux-2.6.16 october2005, patch 05-02
Patch-Mainline: 2.6.18-rc1

  Description: cio: pathgrouping fails due to non-unique pathgroup-IDs
  Symptom:     Grouping of channel paths for access to CCW devices fails under
               some circumstances due to non-unique pathgroup-IDs.
  Problem:     The source for the CPU-ID part of the ID was incorrectly
               specified on 64 bit systems. Additionally, the length of the ID
               was too large due to incorrect data packing declaration.
  Solution:    Fix CPU-ID lowcore address and add missing packing declaration.
  Problem-ID:  24220
  Archs:       s390-31, s390-64

Acked-by: Hannes Reinecke <hare@suse.de>

--- linux-2.5/drivers/s390/cio/css.h	20 Mar 2006 15:57:47 -0000	1.46
+++ linux-2.5/drivers/s390/cio/css.h	30 May 2006 16:09:30 -0000	1.46.2.1
@@ -44,11 +44,11 @@
 	union {
 		__u8 fc;   	/* SPID function code */
 		struct path_state ps;	/* SNID path state */
-	} inf;
+	} __attribute__ ((packed)) inf;
 	union {
 		__u32 cpu_addr	: 16;	/* CPU address */
 		struct extended_cssid ext_cssid;
-	} pgid_high;
+	} __attribute__ ((packed)) pgid_high;
 	__u32 cpu_id	: 24;	/* CPU identification */
 	__u32 cpu_model : 16;	/* CPU model */
 	__u32 tod_high;		/* high word TOD clock */

--- linux-2.5/include/asm-s390/lowcore.h	4 Jan 2006 12:22:38 -0000	1.19
+++ linux-2.5/include/asm-s390/lowcore.h	30 May 2006 16:09:30 -0000	1.19.4.1
@@ -98,8 +98,8 @@
 #define __LC_KERNEL_ASCE		0xD58
 #define __LC_USER_ASCE			0xD60
 #define __LC_PANIC_STACK                0xD68
-#define __LC_CPUID                      0xD90
-#define __LC_CPUADDR                    0xD98
+#define __LC_CPUID                      0xD80
+#define __LC_CPUADDR                    0xD88
 #define __LC_IPLDEV                     0xDB8
 #define __LC_JIFFY_TIMER		0xDC0
 #define __LC_CURRENT			0xDD8

